### PR TITLE
Oracle Delay for Inflation

### PR DESCRIPTION
**Pull request by Joeysantoro**
Allows for the backstop to have a forced delay before triggering, protecting against oracle manipulation attacks or extremely volatile PCV

### CODE REVIEW

File: contracts/stabilizer/TribeReserveStabilizer.sol [hunk 1]
```diff
@@ -55,7 +59,9 @@ contract TribeReserveStabilizer is ITribeReserveStabilizer, ReserveStabilizer {
     /// @notice exchange FEI for minted TRIBE
     /// @dev Collateralization oracle price must be below threshold
     function exchangeFei(uint256 feiAmount) public override returns(uint256) {
-        require(isCollateralizationBelowThreshold(), "TribeReserveStabilizer: Collateralization ratio above threshold");
+        // the timer counts down from first time below threshold
+        // if oracle remains below for entire window, then begin opening exchange
+        require(isTimeEnded(), "TribeReserveStabilizer: Oracle delay");
```
Comment (eswak) [inline contracts/stabilizer/TribeReserveStabilizer.sol:64]: you can also use the `afterTime` modifier I think, but it's clear this way with the comment 👍

File: contracts/stabilizer/TribeReserveStabilizer.sol [hunk 2]
```diff
@@ -72,6 +78,20 @@ contract TribeReserveStabilizer is ITribeReserveStabilizer, ReserveStabilizer {
...
+    /// @notice delay the opening of the TribeReserveStabilizer until oracle delay duration is met
+    function startOracleDelayCountdown() external override {
+        require(isCollateralizationBelowThreshold(), "TribeReserveStabilizer: Collateralization ratio above threshold");
+        require(!isTimeStarted(), "TribeReserveStabilizer: timer started");
+        _initTimed();
+    }
+
+    /// @notice reset the opening of the TribeReserveStabilizer oracle delay as soon as above CR target
+    function resetOracleDelayCountdown() external override {
+        require(!isCollateralizationBelowThreshold(), "TribeReserveStabilizer: Collateralization ratio above threshold");
```
Comment (eswak) [inline contracts/stabilizer/TribeReserveStabilizer.sol:90]: failure message should say _under_ threshold

File: contracts/stabilizer/TribeReserveStabilizer.sol (hunk 1 above)
Comment (Joeysantoro) [reply contracts/stabilizer/TribeReserveStabilizer.sol:64]: Oooh, yeah I like afterTime

File: contracts/utils/Timed.sol [hunk 3]
```diff
@@ -63,6 +63,12 @@ abstract contract Timed {
         emit TimerReset(block.timestamp);
     }
 
+    function _pauseTimer() internal {
```
Comment (ElliotFriedman) [inline contracts/utils/Timed.sol:66]: I don't like adding on to existing contracts that are already deployed. I'd prefer extending the timed class to include this helper method over modifying it.

File: contracts/stabilizer/TribeReserveStabilizer.sol [hunk 4]
```diff
@@ -72,6 +75,20 @@ contract TribeReserveStabilizer is ITribeReserveStabilizer, ReserveStabilizer {
         return valid && ratio.lessThanOrEqualTo(_collateralizationThreshold);
     }
 
+    /// @notice delay the opening of the TribeReserveStabilizer until oracle delay duration is met
+    function startOracleDelayCountdown() external override {
+        require(isCollateralizationBelowThreshold(), "TribeReserveStabilizer: Collateralization ratio above threshold");
+        require(!isTimeStarted(), "TribeReserveStabilizer: timer started");
```
Comment (ElliotFriedman) [inline contracts/stabilizer/TribeReserveStabilizer.sol:81]: Lines 80 and 81 need to have the sad path tested in the unit tests.

**Review by ElliotFriedman — DISMISSED** [review]

File: contracts/stabilizer/TribeReserveStabilizer.sol [hunk 5]
```diff
@@ -72,6 +75,20 @@ contract TribeReserveStabilizer is ITribeReserveStabilizer, ReserveStabilizer {
...
+        _initTimed();
+    }
+
+    /// @notice reset the opening of the TribeReserveStabilizer oracle delay as soon as above CR target
+    function resetOracleDelayCountdown() external override {
+        require(!isCollateralizationBelowThreshold(), "TribeReserveStabilizer: Collateralization ratio under threshold");
```
Comment (ElliotFriedman) [inline contracts/stabilizer/TribeReserveStabilizer.sol:87]: This needs to be unit tested.

File: contracts/stabilizer/TribeReserveStabilizer.sol [hunk 6]
```diff
@@ -72,6 +75,20 @@ contract TribeReserveStabilizer is ITribeReserveStabilizer, ReserveStabilizer {
...
+        require(isTimeStarted(), "TribeReserveStabilizer: timer started");
```
Comment (ElliotFriedman) [inline contracts/stabilizer/TribeReserveStabilizer.sol:88]: same comment

File: contracts/utils/Timed.sol (hunk 3 above)
Comment (ditchfieldcaleb) [reply contracts/utils/Timed.sol:66]: Agreed with @ElliotFriedman here

File: test/unit/stablizer/TribeReserveStabilizer.test.ts [hunk 7]
```diff
@@ -52,7 +54,7 @@ describe('TribeReserveStabilizer', function () {
 
     this.tribeMinter = await (await ethers.getContractFactory('MockTribeMinter')).deploy(this.tribe.address);
```
Comment (ditchfieldcaleb) [inline test/unit/stablizer/TribeReserveStabilizer.test.ts:55]: Should go ahead and remove all instances of `this` in this test

**Review by ditchfieldcaleb — CHANGES_REQUESTED** [review]
Agree with Elliot's comments as well

File: contracts/utils/Timed.sol (hunk 3 above)
Comment (Joeysantoro) [reply contracts/utils/Timed.sol:66]: changed

File: contracts/stabilizer/TribeReserveStabilizer.sol (hunk 4 above)
Comment (Joeysantoro) [reply contracts/stabilizer/TribeReserveStabilizer.sol:80]: done

File: contracts/stabilizer/TribeReserveStabilizer.sol (hunk 5 above)
Comment (Joeysantoro) [reply contracts/stabilizer/TribeReserveStabilizer.sol:86]: done

File: contracts/stabilizer/TribeReserveStabilizer.sol (hunk 6 above)
Comment (Joeysantoro) [reply contracts/stabilizer/TribeReserveStabilizer.sol:87]: done

File: test/integration/tests/backstop.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+});
+
+describe.only('e2e', function () {
+  let contracts: NamedContracts;
+  let contractAddresses: NamedAddresses;
+  let deployAddress: string;
+  let e2eCoord: TestEndtoEndCoordinator;
+  let doLogging: boolean;
+
+  const tenPow18 = toBN('1000000000000000000');
```
Comment (ditchfieldcaleb) [inline test/integration/tests/backstop.ts:27]: Use ethers.constant.WeiPerEther

File: test/integration/tests/backstop.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    doLogging && console.log(`Environment loaded.`);
+  });
+
+  describe('TribeMinter', async function () {
+    it('mint TRIBE', async function () {
+      const { tribeMinter, tribe } = contracts;
+      const tribeSupply = await tribe.totalSupply();
+      const balanceBefore = await tribe.balanceOf(deployAddress);
+
+      await tribeMinter.mint(deployAddress, '100000');
```
Comment (ditchfieldcaleb) [inline test/integration/tests/backstop.ts:56]: Maybe I'm missing something here - but why is the deploy address allowed to mint Tribe for free?

File: test/integration/tests/backstop.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+      await collateralizationOracleWrapper.update();
+
+      const expectedAmountOut = await tribeReserveStabilizer.getAmountOut(feiTokensExchange);
+      await tribeReserveStabilizer.exchangeFei(feiTokensExchange);
+
+      const userFeiBalanceAfter = toBN(await fei.balanceOf(deployAddress));
+      const userTribeBalanceAfter = await tribe.balanceOf(deployAddress);
+
+      expect(userTribeBalanceAfter.sub(toBN(expectedAmountOut))).to.be.equal(userTribeBalanceBefore);
```
Comment (ditchfieldcaleb) [inline test/integration/tests/backstop.ts:89]: I think this test should hand calculated expectedAmountOut; example bug that this wouldn't catch is if expectedAmountOut is zero and it doesn't mint anything.

(either that or unit test expected amount out, which should be unit tested anyway)

File: test/unit/stablizer/TribeReserveStabilizer.test.ts [hunk 11]
```diff
@@ -21,180 +29,196 @@ describe('TribeReserveStabilizer', function () {
...
     it('above', async function () {
-      await this.reserveStabilizer
-        .connect(impersonatedSigners[governorAddress])
-        .setCollateralizationThreshold('9900', {});
-      expect(await this.reserveStabilizer.isCollateralizationBelowThreshold()).to.be.equal(false);
+      await reserveStabilizer.connect(impersonatedSigners[governorAddress]).setCollateralizationThreshold('9900', {});
+      expect(await reserveStabilizer.isCollateralizationBelowThreshold()).to.be.equal(false);
     });
 
     it('at', async function () {
```
Comment (ditchfieldcaleb) [inline test/unit/stablizer/TribeReserveStabilizer.test.ts:182]: use more descripted test language

File: test/unit/stablizer/TribeReserveStabilizer.test.ts [hunk 12]
```diff
@@ -21,180 +29,196 @@ describe('TribeReserveStabilizer', function () {
...
     it('below', async function () {
-      await this.reserveStabilizer
-        .connect(impersonatedSigners[governorAddress])
-        .setCollateralizationThreshold('10000', {});
-      expect(await this.reserveStabilizer.isCollateralizationBelowThreshold()).to.be.equal(true);
+      await reserveStabilizer.connect(impersonatedSigners[governorAddress]).setCollateralizationThreshold('10000', {});
+      expect(await reserveStabilizer.isCollateralizationBelowThreshold()).to.be.equal(true);
     });
 
     it('invalid oracle', async function () {
```
Comment (ditchfieldcaleb) [inline test/unit/stablizer/TribeReserveStabilizer.test.ts:191]: use more descriptive test language, ie "should revert when collateralizationOracle reports invalid status"

**Review by ditchfieldcaleb — DISMISSED** [review]
I think we need more unit tests, maybe with code coverage?

File: test/integration/tests/backstop.ts (hunk 9 above)
Comment (Joeysantoro) [reply test/integration/tests/backstop.ts:56]: Deploy address is a governor in our integration test setup

File: test/integration/tests/backstop.ts (hunk 10 above)
Comment (Joeysantoro) [reply test/integration/tests/backstop.ts:89]: its unit tested in ReserveStabilizer

**Review by ditchfieldcaleb — APPROVED** [review]

**Pull request merged**